Enable wpt WebGL tests on Linux

anholt opened the pull request:
<!-- Please describe your changes on the following line: -->

Enables the WebGL tests in test-wpt on Linux.  They have been disabled since the webgl conformance tests first showed up, and it looks like they should work now.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #9331

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12858)

<!-- Reviewable:end -->

anholt committed:
wpt: Re-enable the webgl tests on Linux.

From a conversation on #servo:

<Manishearth> Ms2ger: why are they disabled on linux only?
<Ms2ger> Don't recall
<Ms2ger> Probably builder issues

Fixes #9331

anholt committed:
wpt: Mark expected Linux webgl panics in DrawingBufferWidth() calls.

anholt committed:
wpt: Mark expected Linux panics due to oom on negative width/height.

It looks like we should basically skip the call of readPixels() in
this case.

anholt committed:
wpt: Downgrade two formerly-crashing tests on Linux to TIMEOUT.

Across many test runs, I haven't seen them crash.

anholt commented:
Note: My testing was done on Mesa's i965 driver on Haswell.  There's an interesting problem in the out-of-bounds tests where with a debug build of Mesa they assertion fail.  I think we may just be getting lucky and not crashing in a release build, requiring better bounds checking on our draws.

Manishearth commented on tests/wpt/metadata/webgl/conformance-1.0.3/conformance/state/gl-geterror.html.ini:
```diff
@@ -2,6 +2,7 @@
   type: testharness
   expected:
     if os == "mac": CRASH
```
this should just be CRASH, since we only test on linux and osx?

Manishearth commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 60a44d8 with merge 12f7c70...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2604)

Manishearth commented:
Seem to be lots of failures on the server :/

anholt committed:
wpt: Mark bindAttribLocation-* test failures on Linux.

anholt committed:
wpt: Disable tex-image-*-canvas*.html that fail on linux-rel

These don't crash on the i965 driver or my local 11.2.2 llvmpipe, but
have shown intermittent failures in CI.

anholt commented:
Yeah, this PR was mostly an attempt to give the CI a shot at it, since its driver would be different from mine.  Hopefully the next force-push will be cleaner on CI.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit acbdc00 with merge 10bf9e5...

anholt commented:
Note the way I added the subtest FAILs in 7ef67c3163018fee1fb04adb113c8200fb354234 -- do I need to file those under linux only somehow?

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2607)

anholt committed:
wpt: Note texture-npot.html failures on CI.

These don't appear on the i965 driver, but seem to be reliable on CI.

anholt commented:
Looks like in round 2 the llvmpipe crashes didn't show up (some spurious failure?) but the texture-npot failures matched.  Update dropped the two crashes and (hopefully correctly) added the texture-npot.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit af99035 with merge 58c909d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2611)

highfive commented:
▶ FAIL [expected PASS] /css-transforms-1_dev/html/transform-table-007.htm
      └   → /css-transforms-1_dev/html/transform-table-007.htm a5c014b20ef1363bea6f24eda28c7efb7c45698a
    /css-transforms-1_dev/html/reference/transform-blank-ref.htm fa6407b1acbbfea27e27061e7d1bdeca98e4a728
    Testing a5c014b20ef1363bea6f24eda28c7efb7c45698a == fa6407b1acbbfea27e27061e7d1bdeca98e4a728

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12857) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 52ab3f4 with merge 08899a2...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2627)

anholt commented:
So, we've got an intermittent fail in the 4444 and 5551 tests in the Mesa driver, and it doesn't appear to be our fault.  Proposed solution here is to disable those two on Linux to keep buildbot complaints down on unrelated PRs, with link to a bug report to track the failure.

emilio commented:
That sounds reasonable.

@bors-servo: r+

Fingers crossed! :)

bors-servo commented:
:pushpin: Commit f8763f4 has been approved by `emilio`

<!-- @bors-servo r=emilio f8763f49d4cf7a8f285e52917b13f6e15de20679 -->

bors-servo commented:
:hourglass: Testing commit f8763f4 with merge 1fb783f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2630)

KiChjang commented:
```
Ran 6839 tests finished in 817.0 seconds.
  • 6838 ran as expected. 1662 tests skipped.
  • 1 tests crashed unexpectedly

Tests with unexpected results:
  ▶ CRASH [expected OK] /webgl/conformance-1.0.3/conformance/textures/tex-image-and-sub-image-2d-with-canvas.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  │ DESCRIPTION: Verify texImage2D and texSubImage2D code paths taking canvas elements (RGBA/UNSIGNED_BYTE)
  │ Testing texImage2D with flipY=true canvas size: 1x2 with red-green
  │ Checking top
  │ Stack trace for thread "WebGLThread"
  │ stack backtrace:
  │    0:     0x7f9de0cd41fd - backtrace::backtrace::trace::h36e5966500ba9bd9
  │    1:     0x7f9de0cd4912 - backtrace::capture::Backtrace::new::h3a3d5e9defd7d407
  │    2:     0x7f9ddf7abb1a - servo::install_crash_handler::handler::h57fd71c5829cfe7c
  │    3:     0x7f9de0947443 - AsmJSFaultHandler
  │                         at /home/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/asmjs/WasmSignalHandlers.cpp:1171
  │    4:     0x7f9ddd18a32f - <unknown>
  │    5:     0x7f9dd2dea805 - _ZN4llvm15ValueHandleBase25AddToExistingUseListAfterEPS0_
  │    6:     0x7f9dd2deb7b9 - 4llvm15ValueHandleBase14ValueIsDeletedEPNS_5Value
  │    7:     0x7f9dd2debe5c - _ZN4llvm5ValueD1Ev
  │    8:     0x7f9dd2dc4506 - <unknown>
  │    9:     0x7f9dd2d27dc4 - _ZN4llvm10BasicBlockD1Ev
  │   10:     0x7f9dd2d27ed8 - _ZN4llvm10BasicBlockD0Ev
  │   11:     0x7f9dd2dac599 - _ZN4llvm8Function17dropAllReferencesEv
  │   12:     0x7f9dd3ff8fc3 - lp_func_delete_body
  │   13:     0x7f9dd3fd63bc - gallivm_jit_function
  │   14:     0x7f9dd3ff619f - draw_llvm_create_variant
  │   15:     0x7f9dd3ff7684 - <unknown>
  │   16:     0x7f9dd3f01ace - <unknown>
  │   17:     0x7f9dd3efaa0e - <unknown>
  │   18:     0x7f9dd3efaede - draw_vbo
  │   19:     0x7f9dd46c5fc1 - <unknown>
  │   20:     0x7f9dd45cd11a - <unknown>
  │   21:     0x7f9dd459feac - <unknown>
  │   22:     0x7f9de0e8e0f5 - webrender_traits::webgl::_<impl webrender_traits..WebGLCommand>::apply::h0d0934d9aca22ca4
  │   23:     0x7f9de0ea56b2 - std::panicking::try::call::hc504b9c10f9fcfe4
  │   24:     0x7f9de126b736 - __rust_maybe_catch_panic
  │   25:     0x7f9de0eac232 - _<F as alloc..boxed..FnBox<A>>::call_box::hfedf8711f9cbad54
  │   26:     0x7f9de12616c2 - std::sys::thread::Thread::new::thread_start::h8f3bd45211e9f5ea
  │   27:     0x7f9ddd182183 - start_thread
  │   28:     0x7f9ddcc9937c - clone
  │   29:                0x0 - <unknown>
  │ Stack trace for thread "WebGLThread"
  │ stack backtrace:
  │    0:     0x7f9de0cd41fd - backtrace::backtrace::trace::h36e5966500ba9bd9
  │    1:     0x7f9de0cd4912 - backtrace::capture::Backtrace::new::h3a3d5e9defd7d407
  │    2:     0x7f9ddf7abb1a - servo::install_crash_handler::handler::h57fd71c5829cfe7c
  │    3:     0x7f9ddcbd5caf - <unknown>
  │    4:     0x7f9ddf7abbab - servo::install_crash_handler::handler::h57fd71c5829cfe7c
  │    5:     0x7f9de0947443 - AsmJSFaultHandler
  │                         at /home/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/asmjs/WasmSignalHandlers.cpp:1171
  │    6:     0x7f9ddd18a32f - <unknown>
  │    7:     0x7f9dd2dea805 - _ZN4llvm15ValueHandleBase25AddToExistingUseListAfterEPS0_
  │    8:     0x7f9dd2deb7b9 - 4llvm15ValueHandleBase14ValueIsDeletedEPNS_5Value
  │    9:     0x7f9dd2debe5c - _ZN4llvm5ValueD1Ev
  │   10:     0x7f9dd2dc4506 - <unknown>
  │   11:     0x7f9dd2d27dc4 - _ZN4llvm10BasicBlockD1Ev
  │   12:     0x7f9dd2d27ed8 - _ZN4llvm10BasicBlockD0Ev
  │   13:     0x7f9dd2dac599 - _ZN4llvm8Function17dropAllReferencesEv
  │   14:     0x7f9dd3ff8fc3 - lp_func_delete_body
  │   15:     0x7f9dd3fd63bc - gallivm_jit_function
  │   16:     0x7f9dd3ff619f - draw_llvm_create_variant
  │   17:     0x7f9dd3ff7684 - <unknown>
  │   18:     0x7f9dd3f01ace - <unknown>
  │   19:     0x7f9dd3efaa0e - <unknown>
  │   20:     0x7f9dd3efaede - draw_vbo
  │   21:     0x7f9dd46c5fc1 - <unknown>
  │   22:     0x7f9dd45cd11a - <unknown>
  │   23:     0x7f9dd459feac - <unknown>
  │   24:     0x7f9de0e8e0f5 - webrender_traits::webgl::_<impl webrender_traits..WebGLCommand>::apply::h0d0934d9aca22ca4
  │   25:     0x7f9de0ea56b2 - std::panicking::try::call::hc504b9c10f9fcfe4
  │   26:     0x7f9de126b736 - __rust_maybe_catch_panic
  │   27:     0x7f9de0eac232 - _<F as alloc..boxed..FnBox<A>>::call_box::hfedf8711f9cbad54
  │   28:     0x7f9de12616c2 - std::sys::thread::Thread::new::thread_start::h8f3bd45211e9f5ea
  │   29:     0x7f9ddd182183 - start_thread
  │   30:     0x7f9ddcc9937c - clone
  └   31:                0x0 - <unknown>
```

anholt commented:
Re-pushed, banning the new intermittent fail, and the preemptively the other one in the same name scheme, too.

emilio commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 7f1c3ca has been approved by `emilio`

<!-- @bors-servo r=emilio 7f1c3ca0ec4ef9a543859623cd5bd6b6e122bea0 -->

bors-servo commented:
:hourglass: Testing commit 7f1c3ca with merge 1c3bcfb...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2633)

anholt commented:
And a new one with **webgl-**canvas this time.  Can I regex these? :(

emilio commented:
I don't think you can, but a quick `find` I guess would do it. Seems like all these are going to be intermittents, judging the stack.

Last frame is `ZN4llvm15ValueHandleBase17RemoveFromUseListEv`, so my guess is that there's a race in llvmpipe when painting at the same time from multiple threads. Seems like they're not locking the `UseList` appropriately. That being said, this is an educated guess and I'll need to verify it, but judging the amount of intermittents and the kind of them...

emilio commented:
[This](https://github.com/llvm-mirror/llvm/blob/2d62ce6ee866bea6f1cd8ab4d9a6c3a6b3801854/lib/IR/Value.cpp#L782) is the code, doesn't seem too suspicious, and doesn't seem by near the reason I was thinking.

@larsbergstrom, could I get access to a builder to debug it?

emilio commented:
Gallium has the option of using a single global context: https://cgit.freedesktop.org/mesa/mesa/tree/src/gallium/drivers/llvmpipe/lp_context.c#n175.

It only seem to be enabled [if compiling for embedded devices](https://cgit.freedesktop.org/mesa/mesa/tree/scons/gallium.py#n359), so I'm not sure that's our case...

That would explain the above stacks, and also why the tests with multiple WebGL canvases running in parallel are flacky. Other tests I think they work fine because in the reftest mode we only issue paint commands once the document is fully loaded, even though for real use cases using llvmpipe wouldn't work.

I've only skimmed the gallium code, does my hypothesis seem reasonable, @anholt?

larsbergstrom commented:
@emilio If you add your SSH key to saltfs (sorta like https://github.com/servo/saltfs/commit/28a263a736f8d2f9edff6bb455ce4715c51d12ac), you will be able to ssh into the builders to try it out.

emilio commented:
Ok, `glxinfo` of the build machine as of right now:

```
root@ip-172-31-14-70:~# DISPLAY=:0 glxinfo
name of display: :0
display: :0  screen: 0
direct rendering: Yes
server glx vendor string: SGI
server glx version string: 1.4
server glx extensions:
    GLX_ARB_multisample, GLX_EXT_import_context, GLX_EXT_texture_from_pixmap, 
    GLX_EXT_visual_info, GLX_EXT_visual_rating, GLX_MESA_copy_sub_buffer, 
    GLX_OML_swap_method, GLX_SGIS_multisample, GLX_SGIX_fbconfig, 
    GLX_SGIX_pbuffer, GLX_SGI_make_current_read
client glx vendor string: Mesa Project and SGI
client glx version string: 1.4
client glx extensions:
    GLX_ARB_create_context, GLX_ARB_create_context_profile, 
    GLX_ARB_create_context_robustness, GLX_ARB_fbconfig_float, 
    GLX_ARB_framebuffer_sRGB, GLX_ARB_get_proc_address, GLX_ARB_multisample, 
    GLX_EXT_create_context_es2_profile, GLX_EXT_fbconfig_packed_float, 
    GLX_EXT_framebuffer_sRGB, GLX_EXT_import_context, 
    GLX_EXT_texture_from_pixmap, GLX_EXT_visual_info, GLX_EXT_visual_rating, 
    GLX_INTEL_swap_event, GLX_MESA_copy_sub_buffer, 
    GLX_MESA_multithread_makecurrent, GLX_MESA_query_renderer, 
    GLX_MESA_swap_control, GLX_OML_swap_method, GLX_OML_sync_control, 
    GLX_SGIS_multisample, GLX_SGIX_fbconfig, GLX_SGIX_pbuffer, 
    GLX_SGIX_visual_select_group, GLX_SGI_make_current_read, 
    GLX_SGI_swap_control, GLX_SGI_video_sync
GLX version: 1.4
GLX extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, GLX_EXT_import_context, 
    GLX_EXT_texture_from_pixmap, GLX_EXT_visual_info, GLX_EXT_visual_rating, 
    GLX_MESA_copy_sub_buffer, GLX_MESA_multithread_makecurrent, 
    GLX_OML_swap_method, GLX_SGIS_multisample, GLX_SGIX_fbconfig, 
    GLX_SGIX_pbuffer, GLX_SGI_make_current_read
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: Gallium 0.4 on llvmpipe (LLVM 3.4, 256 bits)
OpenGL version string: 2.1 Mesa 10.1.3
OpenGL shading language version string: 1.30
```

That means mesa 10.1.3, which is quite old.

emilio commented:
One of the backtraces I got this morning:

```
#0  0x00007fffebc29e6d in ?? () from /usr/lib/x86_64-linux-gnu/libLLVM-3.4.so.1
#1  0x00007fffebc2ab5b in ?? () from /usr/lib/x86_64-linux-gnu/libLLVM-3.4.so.1
#2  0x00007fffeb7d0f77 in llvm::FPPassManager::runOnFunction(llvm::Function&) ()
   from /usr/lib/x86_64-linux-gnu/libLLVM-3.4.so.1
#3  0x00007fffeb7d2f76 in llvm::legacy::FunctionPassManagerImpl::run(llvm::Function&) ()
   from /usr/lib/x86_64-linux-gnu/libLLVM-3.4.so.1
#4  0x00007fffeb7d3044 in llvm::legacy::FunctionPassManager::run(llvm::Function&) ()
   from /usr/lib/x86_64-linux-gnu/libLLVM-3.4.so.1
#5  0x00007fffeb747299 in LLVMRunFunctionPassManager ()
   from /usr/lib/x86_64-linux-gnu/libLLVM-3.4.so.1
#6  0x00007fffec9f50ac in ?? () from /usr/lib/x86_64-linux-gnu/libgallium.so.0
#7  0x00007fffec9f616c in draw_llvm_create_variant ()
   from /usr/lib/x86_64-linux-gnu/libgallium.so.0
#8  0x00007fffec9f7685 in ?? () from /usr/lib/x86_64-linux-gnu/libgallium.so.0
#9  0x00007fffec901acf in ?? () from /usr/lib/x86_64-linux-gnu/libgallium.so.0
#10 0x00007fffec8faa0f in ?? () from /usr/lib/x86_64-linux-gnu/libgallium.so.0
#11 0x00007fffec8faedf in draw_vbo () from /usr/lib/x86_64-linux-gnu/libgallium.so.0
#12 0x00007fffed0c5fc2 in ?? () from /usr/lib/x86_64-linux-gnu/dri/swrast_dri.so
#13 0x00007fffecfcd11b in ?? () from /usr/lib/x86_64-linux-gnu/dri/swrast_dri.so
#14 0x00007fffecf9fead in ?? () from /usr/lib/x86_64-linux-gnu/dri/swrast_dri.so
#15 0x0000555556ff7d87 in layers::rendergl::RenderContext::render_tile::h8550cecef0409ece ()
#16 0x000055555586d3fa in layers::rendergl::RenderContext::render_3d_context::h0c21cdf82368d8fe ()
#17 0x000055555586d72e in layers::rendergl::RenderContext::render_3d_context::h0c21cdf82368d8fe ()
#18 0x000055555586c361 in layers::rendergl::render_scene::hba95fbb4d7ae41f0 ()
#19 0x00005555558635d7 in _$LT$compositing..compositor..IOCompositor$LT$Window$GT$$GT$::composite_specific_target::h15db86bed7dea0f6 ()
#20 0x000055555586bc8f in _$LT$compositing..compositor..IOCompositor$LT$Window$GT$$GT$::composite::hf288dc13a95d85f6 ()
#21 0x00005555558538e0 in _$LT$compositing..compositor..IOCompositor$LT$Window$GT$$GT$::handle_events::h565b6d61b10549df ()
#22 0x0000555555874146 in servo::main::h6924e5658370f6b5 ()
#23 0x000055555733cec8 in std::panicking::try::call::he2cac0ad48ff25a4 ()
#24 0x00005555573455a7 in __rust_maybe_catch_panic ()
#25 0x000055555733c0f3 in std::rt::lang_start::hfe9ab243c60ffb9b ()
#26 0x00007ffff5730f45 in __libc_start_main (main=0x555555876a50 <main>, argc=10, 
    argv=0x7fffffffe3e8, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffe3d8) at libc-start.c:287
#27 0x000055555582cffa in _start ()
```

emilio commented:
I was able to reproduce that consistently inside the debugger, but not outside of it, where the stacks looked just as the previous tests.

Will investigate a bit more when I have the time.

anholt commented:
So what's our status on this?  Should I keep trying to get a set of tests that is stable (if so, a new try would be nice), or @emilio do you want to work on debugging the ancient Mesa?

emilio commented:
I think we're pushing to update mesa (https://github.com/servo/saltfs/issues/468), so it's a bit pointless to debug LLVM3.4 and Mesa 10.

That being said, if we manage to find a stable set of tests, I'm all in with enabling them now.

@bors-servo: try

anholt commented:
I think the try didn't happen?

larsbergstrom commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 8a70c0f with merge 5df6cc2...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2753)

larsbergstrom commented:
Wow, really close! Just 1:
```
  • 1 tests timed out unexpectedly

Tests with unexpected results:
  ▶ TIMEOUT [expected OK] /webgl/conformance-1.0.3/conformance/textures/tex-image-and-sub-image-2d-with-image-rgba4444.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  │ DESCRIPTION: Verify texImage2D and texSubImage2D code paths taking image elements (RGBA/UNSIGNED_SHORT_4_4_4_4)
  │ Testing texImage2D with flipY=true
  └ Checking lower left corner

```

anholt committed:
wpt: Disable a test that failed one in 6 runs on linux-rel.

anholt commented:
OK, marked that one as timeout.  That's a new timeout compared to the other runs, though.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit aa72171 with merge 22742ba...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2761)

highfive commented:
▶ OK [expected TIMEOUT] /webgl/conformance-1.0.3/conformance/textures/tex-image-and-sub-image-2d-with-image-rgba4444.html

anholt commented:
OK, what would people like to see done to this PR at this point?  Disable that test, too?  Drop the TIMEOUT flag on it and file it as an intermittent?

KiChjang commented:
Disable it, since it's clearly and intermittent and we really shouldn't be introducing intermittents when enabling tests.

anholt commented:
OK, pushed with the disable.

emilio commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 549c0c1 has been approved by `emilio`

<!-- @bors-servo r=emilio 549c0c1d67725dfdde0b976085a36807d48074f2 -->

bors-servo commented:
:hourglass: Testing commit 549c0c1 with merge f106f40...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2734)

larsbergstrom commented:
```

ParseError: EOL in expression: /Users/servo/buildbot/slave/mac-rel-wpt/build/tests/wpt/metadata/webgl/conformance-1.0.3/conformance/textures/tex-image-and-sub-image-2d-with-image-rgba4444.html.ini line 3
```

KiChjang commented on tests/wpt/metadata/webgl/conformance-1.0.3/conformance/textures/tex-image-and-sub-image-2d-with-webgl-canvas-rgba4444.html.ini:
```diff
@@ -1,5 +1,7 @@
 [tex-image-and-sub-image-2d-with-webgl-canvas-rgba4444.html]
   type: testharness
+  disabled:
+    if os == "linux": https://github.com/servo/servo/issues/12864
```
Why not disable it for all OSes?

jdm replied:
Because it works on macOS?

anholt replied:
Oh, right.

I was just autopiloting here.  Given that I haven't actually changed anything about MacOS in this series, we should probably drop this commit and ignore the macos failure.

anholt replied:
Scratch that.  The original timeout was linux, this failure was reported on Mac because of the syntax error.

KiChjang commented on tests/wpt/metadata/webgl/conformance-1.0.3/conformance/textures/tex-image-and-sub-image-2d-with-image-rgba4444.html.ini:
```diff
@@ -1,4 +1,6 @@
 [tex-image-and-sub-image-2d-with-image-rgba4444.html]
+  disabled:
+    if os == "linux"
```
And it looks like it's failing because of this line.

KiChjang commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit a0b738d has been approved by `emilio`

<!-- @bors-servo r=emilio a0b738d6b38d0998bac8ef5ae230dd776a382541 -->

bors-servo commented:
:hourglass: Testing commit a0b738d with merge 8491c09...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2736)

anholt commented:
In that last round, the macos timeouts on readpixels tests are unrelated to my changes.

There are also unexpected passes reported on gl-bindAttribLocation-aliasing.html.  I had thought the testsuite was treating the expected-FAIL in subtests as "if the test fails overall, treat these subtests as expected fail and report any ones not listed".  Is that not the case?  Do we need to annotate each line of tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-bindAttribLocation-aliasing.html.ini with "if linux, then expect this"?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13060) made this pull request unmergeable. Please resolve the merge conflicts.

anholt commented:
We've got so many errors different between MacOS and Linux that trying to keep the expecteds updated for both is going to be madness.  I'm shelving this project for now, and instead working on improving our implementation so that we don't have so many expecteds to manage.
